[WebGPU] Pipeline needs to pass zNear and zFar to fragment shader for clamping depth writes

mwyrzykowski opened the pull request:
#### 48520e8acaa0a169267b7e63a4a7f02e6f2c68a3
<pre>
[WebGPU] Pipeline needs to pass zNear and zFar to fragment shader for clamping depth writes
<a href="https://bugs.webkit.org/show_bug.cgi?id=264436">https://bugs.webkit.org/show_bug.cgi?id=264436</a>
&lt;radar://118138930&gt;

Reviewed by Tadeu Zagallo.

WebGPU expects writes to frag_depth to be clipped to the viewport
bounds which is not something which natively exists in Metal, so
we need to pass the viewport bounds to the fragment shader so the
shader can clamp.

* Source/WebGPU/WebGPU/Queue.h:
* Source/WebGPU/WebGPU/Queue.mm:
(WebGPU::Queue::writeBuffer):
* Source/WebGPU/WebGPU/RenderBundle.h:
* Source/WebGPU/WebGPU/RenderBundle.mm:
(WebGPU::RenderBundle::updateMinMaxDepths):
* Source/WebGPU/WebGPU/RenderBundleEncoder.h:
* Source/WebGPU/WebGPU/RenderBundleEncoder.mm:
(-[RenderBundleICBWithResources initWithICB:pipelineState:depthStencilState:cullMode:frontFace:depthClipMode:depthBias:depthBiasSlopeScale:depthBiasClamp:fragmentDynamicOffsetsBuffer:]):
(WebGPU::makeRenderBundleICBWithResources):
(WebGPU::Device::createRenderBundleEncoder):
(WebGPU::RenderBundleEncoder::executePreDrawCommands):
(WebGPU::RenderBundleEncoder::endCurrentICB):
(WebGPU::RenderBundleEncoder::setBindGroup):
(-[RenderBundleICBWithResources initWithICB:pipelineState:depthStencilState:cullMode:frontFace:depthClipMode:depthBias:depthBiasSlopeScale:depthBiasClamp:]): Deleted.
* Source/WebGPU/WebGPU/RenderPassEncoder.h:
* Source/WebGPU/WebGPU/RenderPassEncoder.mm:
(WebGPU::setViewportMinMaxDepthIntoBuffer):
(WebGPU::RenderPassEncoder::executePreDrawCommands):
(WebGPU::RenderPassEncoder::executeBundles):
(WebGPU::RenderPassEncoder::setPipeline):
(WebGPU::RenderPassEncoder::setViewport):
(wgpuRenderPassEncoderExecuteBundles):

Canonical link: <a href="https://commits.webkit.org/270469@main">https://commits.webkit.org/270469@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/84738c859f72a5f61783cb3ea91263d07d55af76

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/25548 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/4153 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/26833 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/27652 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/23416 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/5940 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/1588 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/23563 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/25797 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/3115 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/22025 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/28234 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/2751 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/29071 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/23352 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/23340 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/26913 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/2723 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/975 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/4103 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6138 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/3179 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/3060 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

tadeuzagallo commented on Source/WebGPU/WebGPU/RenderPassEncoder.mm:
```diff
@@ -118,8 +143,14 @@
+    auto fragmentsLastBufferSize = m_fragmentDynamicOffsets.size();
+    ASSERT(fragmentsLastBufferSize >= fragmentValuesAfterDynamicOffsets);
+    using destType = std::remove_reference<decltype(m_fragmentDynamicOffsets[0])>::type;
+    m_fragmentDynamicOffsets[fragmentsLastBufferSize - fragmentValuesAfterDynamicOffsets] = as_type<destType>(m_minDepth);
```
Would WTF::bitwise_cast work here?

mwyrzykowski replied:
Oh nice, I'll use that. I was unaware of it, thank you.

tadeuzagallo commented on Source/WebGPU/WebGPU/RenderPassEncoder.mm:
```diff
@@ -118,8 +143,14 @@
-        [m_renderCommandEncoder setFragmentBytes:&m_fragmentDynamicOffsets[0] length:m_fragmentDynamicOffsets.size() * sizeof(m_fragmentDynamicOffsets[0]) atIndex:m_device->maxBuffersForFragmentStage()];
+    ASSERT(m_fragmentDynamicOffsets.size());
+    static_assert(sizeof(m_fragmentDynamicOffsets[0]) == sizeof(m_minDepth), "expect dynamic offsets container to have matching size to depth values");
+    auto fragmentsLastBufferSize = m_fragmentDynamicOffsets.size();
```
Does this need to be explicitly passed to the compiler?

mwyrzykowski replied:
Wouldn't it be the maximum value of `BindGroupLayoutEntry::fragmentBufferDynamicOffset.has_value() ? *BindGroupLayoutEntry::fragmentBufferDynamicOffset + 3 : 2` in the pipeline layout passed to the compiler?

tadeuzagallo replied:
But frag_depth is an output, which BindGroupLayoutEntry would I use? Maybe I could iterate the list the list of entries and find the highest value?

mwyrzykowski replied:
That would work but maybe it’s simpler for me to pass it since I already have the index. 

I thought initially this wouldn’t work due to auto layout, but it is not possible to use dynamic offsets with auto layout. So I would just pass 2 in that case I think.

mwyrzykowski committed:
[WebGPU] Pipeline needs to pass zNear and zFar to fragment shader for clamping depth writes
https://bugs.webkit.org/show_bug.cgi?id=264436
<radar://118138930>

Reviewed by Tadeu Zagallo.

WebGPU expects writes to frag_depth to be clipped to the viewport
bounds which is not something which natively exists in Metal, so
we need to pass the viewport bounds to the fragment shader so the
shader can clamp.

* Source/WebGPU/WebGPU/Queue.h:
* Source/WebGPU/WebGPU/Queue.mm:
(WebGPU::Queue::writeBuffer):
* Source/WebGPU/WebGPU/RenderBundle.h:
* Source/WebGPU/WebGPU/RenderBundle.mm:
(WebGPU::RenderBundle::updateMinMaxDepths):
* Source/WebGPU/WebGPU/RenderBundleEncoder.h:
* Source/WebGPU/WebGPU/RenderBundleEncoder.mm:
(-[RenderBundleICBWithResources initWithICB:pipelineState:depthStencilState:cullMode:frontFace:depthClipMode:depthBias:depthBiasSlopeScale:depthBiasClamp:fragmentDynamicOffsetsBuffer:]):
(WebGPU::makeRenderBundleICBWithResources):
(WebGPU::Device::createRenderBundleEncoder):
(WebGPU::RenderBundleEncoder::executePreDrawCommands):
(WebGPU::RenderBundleEncoder::endCurrentICB):
(WebGPU::RenderBundleEncoder::setBindGroup):
(-[RenderBundleICBWithResources initWithICB:pipelineState:depthStencilState:cullMode:frontFace:depthClipMode:depthBias:depthBiasSlopeScale:depthBiasClamp:]): Deleted.
* Source/WebGPU/WebGPU/RenderPassEncoder.h:
* Source/WebGPU/WebGPU/RenderPassEncoder.mm:
(WebGPU::setViewportMinMaxDepthIntoBuffer):
(WebGPU::RenderPassEncoder::executePreDrawCommands):
(WebGPU::RenderPassEncoder::executeBundles):
(WebGPU::RenderPassEncoder::setPipeline):
(WebGPU::RenderPassEncoder::setViewport):
(wgpuRenderPassEncoderExecuteBundles):

Canonical link: https://commits.webkit.org/270469@main
